### PR TITLE
plot.line(): Draw multiple lines for 2D DataArrays.

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Tests added (for all bug fixes or enhancements)
 - [x] Tests passed (for all non-documentation changes)
 - [x] Passes ``git diff upstream/master **/*py | flake8 --diff`` (remove if you did not edit any Python files)
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API (remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later)

Adds support for plotting multiple lines if `plot.line()` is provided with a 2D dataarray.

kwarg `x` lets you specify co-ordinate for x-axis.

Example:

xarray-multiple-line

### CODE REVIEW

**Comment by dcherian** [comment]
Now adds a legend too. Can be turned off with `add_legend=False`

xarray-multiple-line

**Comment by fmaussion** [comment]
Looks quite good! Thanks. 

Sorry to be annoying ;), but there should be tests for the features you document:

> +    x : string, optional
> +        Coordinate for x axis (2D inputs only). If None use darray.dims[1]
> +    add_legend : boolean, optional
> +        Add legend with y axis coordinates (2D inputs only).

Currently, choosing another dim for the x axis and the legend kwargs are not tested.

**Comment by dcherian** [comment]
@fmaussion I've added more tests. Let me know if you'd like more changes.

**Comment by shoyer** [comment]
Is there a keyword argument name that it would make sense to use for the dimension that is repeated into multiple lines? Maybe `hue`, as is used by Seaborn?

I would like there to be a fully explicit way to make these plots.

**Comment by dcherian** [comment]
@fmaussion I've added one more commit. If `x` kwarg is not provided, we automatically choose the longer dimension. So if you have a 10000x3 DataArray, this will plot 3 lines instead of 10000.

Extended an existing test

**Comment by dcherian** [comment]
@shoyer `hue` sounds perfect. I'll work on adding that.

**Comment by shoyer** [comment]
>  If x kwarg is not provided, we automatically choose the longer dimension. So if you have a 10000x3 DataArray, this will plot 3 lines instead of 10000.

I am very nervous about automated heuristics for choosing behavior. I would much rather we raise an error message in cases like this, rather than guessing. (The problem is that heuristics can make it very hard to predict/understand how code will work without trying it.)

**Comment by dcherian** [comment]
I've added support for a `hue` kwarg in the latest commit. Let me know what you think.

> I am very nervous about automated heuristics for choosing behavior. I would much rather we raise an error message in cases like this, rather than guessing. (The problem is that heuristics can make it very hard to predict/understand how code will work without trying it.)

Well, this behaviour is analogous to automatically choosing `x,y` in `_plot2d`. xarray is already pretty opinionated in this respect.

The current behaviour for a 10000x3 array is to plot 10000 lines which is bad. I'm OK with adding an error message but strongly feel that choosing to plot 3 lines (i.e. always smallest number of lines) is a good default. 

Re:error, would that be a message stating that either `x` or `hue` be specified if the input is 2D?

**Comment by shoyer** [comment]
> Well, this behaviour is analogous to automatically choosing x,y in _plot2d. xarray is already pretty opinionated in this respect.

We use the order of the dimensions on the array for choosing how to plot it. The analogous behavior would be to always plot longer dimension along the x-axis, which isn't what we do.

> The current behaviour for a 10000x3 array is to plot 10000 lines which is bad. I'm OK with adding an error message but strongly feel that choosing to plot 3 lines (i.e. always smallest number of lines) is a good default.

I agree that users probably rarely want 10,000 lines :). That's a good reason to require an explicit choice here.

The problem are edge cases like a 5x6 array. Do you want 5 lines of 6 points each or 6 lines of 5 points each? If we make the heuristic depend on the size of the array, then it will give very hard to understand what happens when the array shape changes slightly.

> Re:error, would that be a message stating that either x or hue be specified if the input is 2D?

Yes, that sounds right.

**Comment by dcherian** [comment]
@shoyer , @fmaussion : Updated to require either `x` or `hue` kwarg if input is 2D.

File: xarray/plot/plot.py [hunk 1]
```diff
@@ -193,11 +199,26 @@ def line(darray, *args, **kwargs):
...
+    add_legend = kwargs.pop('add_legend', True)
 
     ax = get_axis(figsize, size, aspect, ax)
 
-    xlabel, = darray.dims
-    x = darray.coords[xlabel]
+    if ndims == 1:
+        xlabel, = darray.dims
+        if x is not None and xlabel != x:
+            raise ValueError('Input does not have specified dimension ' + x)
```
Comment (shoyer) [inline xarray/plot/plot.py]: please use string formatting instead so this still works even if the dimensions is not a string, e.g.,
```
raise ValueError('Input does not have specified dimension {!r}'.format(x))
```
(Using non-strings as dimension names is rare but we still try to support it when feasible.)

**Comment by dcherian** [comment]
@shoyer Used your suggested change.

File: xarray/plot/plot.py [hunk 2]
```diff
@@ -208,7 +208,8 @@ def line(darray, *args, **kwargs):
     if ndims == 1:
         xlabel, = darray.dims
         if x is not None and xlabel != x:
-            raise ValueError('Input does not have specified dimension ' + x)
+            raise ValueError('Input does not have specified dimension'
+                             + ' {!r}'.format(x))
```
Comment (shoyer) [inline xarray/plot/plot.py]: just for future reference: you can skip the `+` character and rely on implicit string concatenation.

Comment (dcherian) [reply xarray/plot/plot.py]: Didn't know about that. Thanks.

**Comment by shoyer** [comment]
Thanks!

**Comment by shoyer** [comment]
I merged this, but it might also be nice to add an example (or at least brief mention) to the narrative docs for 1D plotting. See:
http://xarray.pydata.org/en/stable/plotting.html#one-dimension
https://github.com/pydata/xarray/blob/master/doc/plotting.rst

**Pull request merged**